File: example-android-views/app/src/main/java/co/verifyspeed/example/android/views/ui/main/MainViewModel.java
package co.verifyspeed.example.android.views.ui.main;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import co.verifyspeed.example.android.views.data.VerifySpeedService;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MainViewModel extends AndroidViewModel {
    private final VerifySpeedService verifySpeedService;
    private final MutableLiveData<String> countryCode = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<String> phoneNumber = new MutableLiveData<>();
    private final MutableLiveData<Boolean> showSuccessDialog = new MutableLiveData<>(false);
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    public MainViewModel(Application application) {
        super(application);
        verifySpeedService = new VerifySpeedService();
        
        // Initialize country code
        executorService.execute(() -> {
            try {
                VerifySpeedService.Result<String> result = verifySpeedService.getCountry().get();
                if (result.isSuccess()) {
                    countryCode.postValue(result.getValue());
                } else {
                    error.postValue(result.getError().getMessage());
                }
            } catch (Exception e) {
                error.postValue(e.getMessage());
            }
        });
    }

    public LiveData<String> getCountryCode() {
        return countryCode;
    }

    public LiveData<Boolean> getIsLoading() {
        return isLoading;
    }

    public LiveData<String> getError() {
        return error;
    }

    public LiveData<String> getPhoneNumber() {
        return phoneNumber;
    }

    public LiveData<Boolean> getShowSuccessDialog() {
        return showSuccessDialog;
    }

    public void showSuccessDialog() {
        showSuccessDialog.setValue(true);
    }

    public void hideSuccessDialog() {
        showSuccessDialog.setValue(false);
    }

    public void getPhoneNumberFromToken(String token) {
        isLoading.setValue(true);
        executorService.execute(() -> {
            try {
                String number = verifySpeedService.getPhoneNumber(token).get();
                phoneNumber.postValue(number);
            } catch (Exception e) {
                error.postValue(e.getMessage());
            } finally {
                isLoading.postValue(false);
            }
        });
    }

    public void setPhoneNumber(String number) {
        phoneNumber.setValue(number);
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        executorService.shutdown();
    }
}

File: example/app/src/main/java/co/verifyspeed/example/ui/main/MainViewModel.kt
package co.verifyspeed.example.ui.main

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import co.verifyspeed.example.data.VerifySpeedService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MainViewModel(application: Application) : AndroidViewModel(application) {
    val verifySpeedService = VerifySpeedService()
    private val _countryCode = MutableStateFlow<String?>(null)
    val countryCode = _countryCode.asStateFlow()
    var isLoading by mutableStateOf(false)
        private set
    var error by mutableStateOf<String?>(null)
        private set
    var phoneNumber by mutableStateOf<String?>(null)
        private set
    var showSuccessDialog by mutableStateOf(false)
        private set

    init {
        viewModelScope.launch {
            verifySpeedService.getCountry().onSuccess { code -> _countryCode.value = code }
        }
    }

    fun showSuccessDialog() {
        showSuccessDialog = true
    }

    fun hideSuccessDialog() {
        showSuccessDialog = false
    }

    suspend fun getPhoneNumberFromToken(token: String) {
        try {
            isLoading = true
            val number = verifySpeedService.getPhoneNumber(token)
            phoneNumber = number
        } catch (e: Exception) {
            // Handle error if needed
        } finally {
            isLoading = false
        }
    }
}
